Add a Count-Min-based probabilistic relative frequency sketch

vekterli opened the pull request:
@havardpe please review. This is arguably a rather experimental addition (that is currently not wired to anything), but it could come in handy for gluing into the caching subsystem as part of a cache admission controller (or something else shiny) 😳👉👈

Adds an implementation of a probabilistic frequency sketch that allows for estimating the relative frequency of of elements from a stream of events. That is, the sketch does not capture the _absolute_ frequency of a given element over time.

To reduce the requirement for the number of bits used for the sketch's underlying counters, this sketch uses automatic decaying of counter values once the number of recorded samples reaches a certain point (relative to the sketch's size). Decaying divides all counters by 2.

The underlying data structure is a Count-Min sketch [^0][^1] with automatic decaying of counters based on TinyLFU [^2].

This implementation has certain changes from a "textbook" CM sketch, inspired by the approach used in [^3]. In particular, instead of having `d` logical rows each with width `w` that are accessed with hash-derived indexes (and thus likely triggering `d` cache misses for large values of `w`) we subdivide into `w/64` blocks each with fixed number d=4 rows of 32 4-bit counters, i.e. each block is exactly 64 bytes. Counter updates or reads always happen within the scope of a single block. We also ensure the block array is allocated with at least a 64-byte alignment. This ensures that a given sketch update will touch exactly 1 cache line of the underlying sketch buffer (not counting cache lines occupied by the sketch object itself, as we assume these are already present in the CPU cache). Similarly, comparing the frequency of two elements will always touch at most 2 cache lines.

The Count-Min sketch (and its cousin, the Counting Bloom Filter) using `k` counters is usually described as requiring `k` pairwise independent hash functions. This implementation assumes this requirement is unnecessary assuming a hash function with good entropy; we instead extract non-overlapping subsets of bits of a single 64-bit hash value and use these as indices into our data structure components.

 [^0]: The Count-Min Sketch and its Applications (2003)
 [^1]: https://en.wikipedia.org/wiki/Count%E2%80%93min_sketch
 [^2]: TinyLFU: A Highly Efficient Cache Admission Policy (2015)
 [^3]: [Caffeine FrequencySketch](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/FrequencySketch.java)

vekterli committed:
Add a Count-Min-based probabilistic relative frequency sketch

Adds an implementation of a probabilistic frequency sketch that allows
for estimating the relative frequency of of elements from a stream of events.
That is, the sketch does not capture the _absolute_ frequency of a given
element over time.

To reduce the requirement for the number of bits used for the sketch's
underlying counters, this sketch uses automatic decaying of counter values
once the number of recorded samples reaches a certain point (relative to
the sketch's size). Decaying divides all counters by 2.

The underlying data structure is a Count-Min sketch [0][1] with automatic
decaying of counters based on TinyLFU [2].

This implementation has certain changes from a "textbook" CM sketch,
inspired by the approach used in [3]. In particular, instead of having `d`
logical rows each with width `w` that are accessed with hash-derived
indexes (and thus likely triggering `d` cache misses for large values
of `w`) we subdivide into w/64 blocks each with fixed number d=4 rows of
32 4-bit counters, i.e. each block is exactly 64 bytes. Counter updates or
reads always happen within the scope of a single block. We also ensure the
block array is allocated with at least a 64-byte alignment. This ensures
that a given sketch update will touch exactly 1 cache line of the
underlying sketch buffer (not counting cache lines occupied by the sketch
object itself, as we assume these are already present in the cache).
Similarly, comparing the frequency of two elements will always touch at
most 2 cache lines.

The Count-Min sketch (and its cousin, the Counting Bloom Filter) using `k`
counters is usually described as requiring `k` pairwise independent hash
functions. This implementation assumes this requirement is unnecessary
assuming a hash function with good entropy; we instead extract
non-overlapping subsets of bits of a single hash value and use these as
indices into our data structure components.

References:
 [0]: The Count-Min Sketch and its Applications (2003)
 [1]: https://en.wikipedia.org/wiki/Count%E2%80%93min_sketch
 [2]: TinyLFU: A Highly Efficient Cache Admission Policy (2015)
 [3]: https://github.com/ben-manes/caffeine/blob/master/caffeine/
      src/main/java/com/github/benmanes/caffeine/cache/FrequencySketch.java

ben-manes commented:
fwiw, I recently added the sketch benchmarks to CI and the results were opposite of what I recall. For some reason the previous version, which was not optimized for a single cache line, was faster in those [results](https://jmh.morethan.io/?gists=f93df1b66aa210bc23a1e672ef9bee16). I then saw the same on my laptop.

The [commit](https://github.com/ben-manes/caffeine/commit/4e714756230c705b6988bd3af29fbe00a4a537fc) introducing this said it was 2.4-2.5x faster. I haven't reverted to benchmark at that commit and see if there was a regression.

Strangely this was not only my observation of the block sketch being faster. The improvement was ported to C# ([results](https://github.com/bitfaster/BitFaster.Caching/pull/271)) and Go ([results](https://github.com/Yiling-J/theine-go/pull/10#issuecomment-1532877652)).

I plan on revisiting this once Java supports a [SIMD API](https://openjdk.org/jeps/489), since that would justify optimizations that could lead to some rework. The performance difference wasn't a problem originally or now, making it mostly done for fun, due we use the sketch. If your usage is more sensitive then you might dig into it, as I'm unsure what benchmark results are accurate at the moment.

havardpe commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.h:
```diff
@@ -0,0 +1,134 @@
+public:
+    // Increments the estimated frequency for the given element, identified by its hash.
+    // Frequency is saturated at 15.
+    void add(const T& elem) noexcept {
```
consider returning the updated min_count here, to avoid having to re-hash the entry to compare with the worst element in the cache later. This would make the estimate_relative_frequency function redundant

vekterli replied:
I've added a separate `add_and_count` method with these semantics—the existing `add` remains `void`. This is because I noticed the function epilogue gets a chain of `CMOV` instructions when extracting the minimum counter value, and I prefer not to include these unless required.

Both functions are simple forwards to a `template <bool ReturnMinCount> uint8_t add_by_hash_impl` function.

[Godbolt example](https://godbolt.org/z/78scd8vY3) (please casually ignore the ugly bunch of added function arguments that emulate `this` 🙈). Also demonstrates the auto-vectorization of the counter block division. Clang in particular turns this into a 256-bytes-at-a-time beast.

havardpe reviewed (commented):
Looks good. I have some minor comments

vekterli committed:
Address review feedback

 * Only use 5 bits from hash per counter selection.
 * Add distinct `concept` for sketch hasher.
 * Add separate `add_and_count` method that combines updatating
   frequency and returning the new Min-Count.
 * Remove heuristic for _not_ counting sample towards decay. Might
   be re-added later once we have a realistic data set to test with.

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.h:
```diff
@@ -0,0 +1,134 @@
+ * See `RawRelativeFrequencySketch` for algorithm details.
+ */
+template <typename T, typename Hash = std::hash<T>, bool HasGoodEntropyHash = false>
+requires requires(Hash h, T t) { noexcept(noexcept(h(t))); }
```
Fixed—moved to distinct `SketchHasher` concept.

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.h:
```diff
@@ -0,0 +1,134 @@
+requires requires(Hash h, T t) { noexcept(noexcept(h(t))); }
+class RelativeFrequencySketch {
+    RawRelativeFrequencySketch _impl;
+    Hash                       _hash;
```
Done

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.h:
```diff
@@ -0,0 +1,134 @@
+        return _impl.estimate_relative_frequency_by_hash(lhs_hash, rhs_hash);
+    }
+    // Manually trigger counter decay; divides all count estimates by 2
+    void div_all_by_2() {
```
Leftover from manual testing; removed from non-raw sketch and moved raw sketch impl to `private`

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.cpp:
```diff
@@ -0,0 +1,161 @@
+            // Compiler will optimize away memcpys (avoids aliasing).
+            memcpy(&chunk, block_ptr + (sizeof(uint64_t) * j), sizeof(uint64_t));
+            chunk >>= 1;
+            chunk &= 0x7777777777777777ULL; // nibble ~MSB mask
```
Done

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.cpp:
```diff
@@ -0,0 +1,161 @@
+        for (uint32_t j = 0; j < 8; ++j) {
+            uint64_t chunk;
+            // Compiler will optimize away memcpys (avoids aliasing).
+            memcpy(&chunk, block_ptr + (sizeof(uint64_t) * j), sizeof(uint64_t));
```
Done

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.cpp:
```diff
@@ -0,0 +1,161 @@
+    const uint8_t* block_ptr = static_cast<const uint8_t*>(_buf.get()) + (block * 64);
+    uint8_t cm[4];
+    for (uint8_t i = 0; i < 4; ++i) {
+        uint8_t h = hash >> (i*8);
```
This was intentional as a way to have a power of two multiplication, but in retrospect it reduces the maximum number of potential blocks too much. Moved to explicitly 5 bits instead.

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.cpp:
```diff
@@ -0,0 +1,161 @@
+      _window_size((_buf.size() / 8) * 10),
+      _block_mask_bits(_buf.size() > 64 ? Optimized::msbIdx(_buf.size() / 64) : 0)
+{
+    assert(_block_mask_bits <= 48); // Will always be the case in practice, but it's an invariant...
```
Good catch, using 44 since we now use 5 bits

vekterli commented on vespalib/src/vespa/vespalib/util/relative_frequency_sketch.cpp:
```diff
@@ -0,0 +1,161 @@
+ */
+RawRelativeFrequencySketch::RawRelativeFrequencySketch(size_t count)
+    : _buf(alloc::Alloc::alloc_aligned(roundUp2inN(std::max(size_t(64U), count * 8)), 512)),
+      _samples_since_decay(0),
```
Renamed to `_estimated_sample_count`

vekterli commented:
@havardpe PTAL

ben-manes commented:
The summary is hard to read since it is comparing across Java versions. Instead click on the result json link above for an individual bar chart. I refer to the [old version](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/jmh/java/com/github/benmanes/caffeine/cache/sketch/CountMinSketch.java) as “flat” for lack of a better term.

My guess is that it’s Java specific and maybe I broke escape analysis. To avoid data dependencies for instruction parallelism, I accumulate the results into an array, assuming the JIT will loop unroll, etc. An array is heap allocated, like all objects, but since it never escapes the method it should be optimized to the stack, etc. If that’s no longer happening then allocations add memory pressure and limit performance.

It’s really hard to rely on implicit optimizations that you cannot force explicitly, like I can by using explicit shifts instead of power-by-2 multiplication. I might have broken some delicate structure that helped the compiler realize it could apply these types of improvements, but that’s very difficult to design or test for. I’m assuming it’s my mistake and haven’t investigated yet, so it’s just an fyi of an unexpected change.

vekterli commented:
Understood, thank you very much for the heads up!

The lack of a-priori insight into what the JVM is going to do with a piece of performance-critical code is a not-too-infrequent headache, which I will readily whine about to anyone unfortunate enough to be within whining-distance. Godbolt can be used with Java, but it only shows the JVM byte code—which is of dubious value.

_The JIT giveth, and the JIT taketh away..._

ben-manes commented:
I took a quick glance at the benchmark and realized a subtle refactoring might explain part of it. I originally benchmarked by simply converting the `FrequencySketch` between the two versions for side-by-side runs. When I moved it into a longer-term commit and did later clean ups, I copied the old one into the test code, eagerly initialized its capacity, removed the initialization "if" guard checks on the methods, and had it implement a common interface (while using a proxy for Caffeine's actual version). That was enough to skew since we are talking about instruction-level parallelism, branch predictions, inlining, etc in the final output. What I should have done is kept it identical so there would be no unfair compiler advantage.

That now shows an improvement for large sketch sizes due to the single memory access. It is not as pronounced as 2x better and at smaller sizes it is still slower. So probably more when I have time to dig into it, and simply stole a little during a work meeting when my mind drifted...

ben-manes commented:
Removing the `count[]` and `index[]` arrays fixed the remaining issues so at worst case the block-based has equal performance. Ideally these should have been stack allocated and easily optimized by the compiler into their individual components. The intent was to avoid data dependencies so that the loop would be independent and trivially unrolled. The `Math.min` is a branch-free single cycle intrinsic, so despite the stall it is easy for a compiler and OOO pipeline to optimize. For `increment`, manually loop unrolling (like the flat version does) added another 10M ops/s to close the gap at the smallest size and increase its lead at the largest.

I'm still confused as to why benchmarks results differ so much from my original analysis or those by independent ports. I'm probably still missing something or was foolish before, since the gains are smaller and I did try to benchmark each optimization idea. If I restore to the original commit there is only a +/- 10M ops/s difference that favors flat on small tables and block on large. That's not enough to have justified a change, nor explain why others saw it, but perhaps the older machine (M3 Max now, 2016 Intel then) caused some skew... 
 
Anyway, sorry for the noise but it was confusing me, I saw your changes, and felt a need to sort it out.

ben-manes commented:
Oh, the last remaining issue is simply the extra multiplication latency for a small table. Since its 128kb and the benchmark uses zipf distribution to mimic hot/cold cache access, this mostly stays in L1. The flat sketch does a 2 round hash and each loop does its own 3rd round for the index, whereas the block-based does a 3rd round prior to the loop to calculate all indexes. That data dependency means the cost of the multiplication in a microbenchmark is visible, whereas in the flat is parallelized so the cost is amortized to be cheaper. One could use a bitwise mixing operations instead, but this is a bit ridiculous as unrelated to visible performance and merely serves to explains the benchmark results for a small table.

vekterli commented:
Interesting! I don't have any instruction timings for the M-series CPUs, but at least on Skylake-era x64 a `MUL` looks like it has a 3-4 cycle latency, so although it's not much it's still time spent twiddling thumbs in the pipeline.

This implementation by default assumes that the provided hash function has poor entropy (which is more often than not the case with standard library hashes—these tend to be the identity function for integers 😬) and mixes it up with XXH3 once prior to use. Since this is a full 64-bit hash of (presumably) good entropy we have enough distinct bits to feed to both the block and counter index calculations.

Hmm, I suppose another difference between x64 and the M-series is 128-byte cache lines. A `long` array with unfortunate alignment risks pulling in (and for increments—updating) 2 cache lines of 128 bytes each for half the addressable blocks. But I would expect that large cache lines should affect the flat table even more...

ben-manes commented:
I think 64 bytes is still the standard cache line size, but often 2-4 lines are pulled in at once if there is no MESI restrictions. The spatial prefetcher guesses that locality might avoid compulsory misses, today’s caches are large, and it’s cheaper to waste space than wait for a memory access. They try to be aggressive but also back off to avoid bandwidth saturations and false sharing problems. I’m guessing that my older 2016 intel was a lot simpler with its smaller caches so it probably made the block look better, whereas current hardware is so over provisioned that it’s harder to observe the benefits. Regardless it’s all for fun and it doesn’t matter what we do in practice, it’s already fast enough.
